File: Extensions/Tessa.Extensions.Default.Client/Workflow/KrProcess/KrLocalTileCommand.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Localization;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tiles;

namespace Tessa.Extensions.Default.Client.Workflow.KrProcess
{
    /// <summary>
    /// Команда выполняющаяся при клике по локальному тайлу подсистемы маршрутов.
    /// </summary>
    /// <param name="launcher"><inheritdoc cref="IKrProcessLauncher" path="/summary"/></param>
    /// <param name="krSecondaryProcessTileUIHandlerResolver"><inheritdoc cref="IKrSecondaryProcessTileUIHandlerResolver" path="/summary"/></param>
    public sealed class KrLocalTileCommand(
        IKrProcessLauncher launcher,
        IKrSecondaryProcessTileUIHandlerResolver krSecondaryProcessTileUIHandlerResolver) :
        IKrTileCommand
    {
        #region Fields

        private readonly IKrProcessLauncher launcher = NotNullOrThrow(launcher);
        private readonly IKrSecondaryProcessTileUIHandlerResolver krSecondaryProcessTileUIHandlerResolver = NotNullOrThrow(krSecondaryProcessTileUIHandlerResolver);

        #endregion

        #region IKrTileCommand Members

        /// <inheritdoc />
        public async Task OnClickAsync(
            IUIContext context,
            ITile tile,
            KrTileInfo tileInfo)
        {
            if (tileInfo.ID == Guid.Empty
                || context.CardEditor is not { } cardEditor)
            {
                return;
            }

            Dictionary<string, object?>? additionalInfo = null;

            if (tileInfo.HandlerID is not null
                && this.krSecondaryProcessTileUIHandlerResolver.TryResolve(tileInfo.HandlerID.Value) is { } handler
                && !await handler.HandleTileAsync(
                        context,
                        tileInfo,
                        additionalInfo = [],
                        CancellationToken.None))
            {
                return;
            }

            if (tileInfo.AskConfirmation
                && !TessaDialog.Confirm(LocalizeFormat(tileInfo.ConfirmationMessage)))
            {
                return;
            }

            using (TessaSplash.Create("$KrButton_DefaultTileSplash"))
            using (cardEditor.SetOperationInProgress(blocking: true))
            {
                // Двойное сохранение, если в карточке есть изменения. Сперва делается основное сохранение, затем запуск процесса.
                if (await cardEditor.CardModel.HasChangesAsync())
                {
                    if (!await cardEditor.SaveCardAsync(
                            context,
                            request: new CardSavingRequest(CardSavingMode.RefreshOnSuccess)))
                    {
                        return;
                    }
                }

                var process = KrProcessBuilder
                    .CreateProcess()
                    .SetProcess(tileInfo.ID)
                    .SetCard(context.CardEditor.CardModel.Card.ID)
                    .SetProcessInfo(additionalInfo)
                    .Build();

                await this.launcher.LaunchWithCardEditorAsync(
                    process,
                    cardEditor,
                    true);
            }
        }

        #endregion
    }
}

File: WebClient SDK/src/default/workflow/krProcess/krLocalTileCommand.ts
import { IKrTileCommand } from './interfaces';
import { IUIContext, showConfirm } from 'tessa/ui';
import {
  KrSecondaryProcessTileUIHandlerRegistry,
  launchProcessWithCardEditor
} from 'tessa/ui/workflow/krProcess';
import { Guid } from 'tessa/platform';
import { ICardModel, CardSavingRequest, CardSavingMode } from 'tessa/ui/cards';
import { ITile } from 'tessa/ui/tiles';
import { KrProcessInstance, KrTileInfo } from '@tessa/platform';
import { IStorage } from '@tessa/core';

export class KrLocalTileCommand implements IKrTileCommand {
  //#region ctor

  private constructor() {}

  //#endregion

  //#region instance

  private static _instance: KrLocalTileCommand;

  public static get instance(): KrLocalTileCommand {
    if (!KrLocalTileCommand._instance) {
      KrLocalTileCommand._instance = new KrLocalTileCommand();
    }
    return KrLocalTileCommand._instance;
  }

  //#endregion

  //#region IKrTileCommand members

  public async onClickAction(
    context: IUIContext,
    _tile: ITile | null,
    tileInfo: KrTileInfo
  ): Promise<void> {
    const cardEditor = context.cardEditor;
    let model: ICardModel;
    if (tileInfo.id === Guid.empty || !cardEditor || !(model = cardEditor.cardModel!)) {
      return;
    }

    let additionalStorage: IStorage | undefined = undefined;
    if (tileInfo.handlerId) {
      additionalStorage = {};
      const handler = KrSecondaryProcessTileUIHandlerRegistry.instance.get(tileInfo.handlerId);
      if (handler && !(await handler.handleTile(context, tileInfo, additionalStorage))) {
        return;
      }
    }

    if (tileInfo.askConfirmation) {
      const result = await showConfirm(tileInfo.confirmationMessage);
      if (!result) {
        return;
      }
    }

    await cardEditor.setOperationInProgress(async () => {
      if (await model.hasChanges()) {
        if (
          !(await cardEditor.saveCard(
            context,
            undefined,
            new CardSavingRequest(CardSavingMode.RefreshOnSuccess)
          ))
        ) {
          return;
        }
      }
      const process = KrProcessInstance.create({
        processId: tileInfo.id,
        processInfo: additionalStorage,
        cardId: context.cardEditor!.cardModel!.card.id
      });

      await launchProcessWithCardEditor(process, cardEditor, true);
    });
  }

  //#endregion
}
